README.md
# B2B Trading Platform



CHANGELOG.md

0.1.6 / 2017-01-24
==================

  * Merge branch 'bug/Issue7': Add support for standalone node.
  * Add git pre-commit hook to lint and test
  * Fix `NODE_PATH`.
  * By default, enable SSR in production mode.
  * Revert "Replace `NODE_PATH` with `app-module-path`"
  * Fix minor issue.
  * Make `HOST` an env to easily use `localhost`;
  * For now, turn off SSR by default.

0.1.5 / 2017-01-23
==================

  * Improve SSR.

0.1.4 / 2017-01-23
==================

  * Improve login form.
  * Add profile info to header.
  * Icon component.
  * Don't let body overflow.

0.1.3 / 2017-01-18
==================

  * Improved scrolling.
  * Simplify graphql loader hook.
  * Implement webpack tree shaking.
  * Fix offline plugin cache error in production mode.
  * Correctly minify html.
  * Delay load animation

0.1.2 / 2017-01-15
==================

  * [EXPERIMENTAL]: Add support for SSR.
  * Update failing tests.

0.1.1 / 2017-01-13
==================

  * Site notifications, many improvements and bug fixes.
  * Partially resolve Issue #4.
  * Improve sign up process.
  * Improve linting.
  * Add notification after resending email verification.
  * Partially resolve Issue #4.
  * Fix: Issue #3.
  * Use snackbar in password reset.
  * Revamp snackbar. Prepare for site notifications.
  * Add immutable proptypes.
  * Small fix.
  * Small fix.
  * Dedupe style imports.
  * Update bootstrap.
  * A few optimizations.
  * Update deps.

0.1.0 / 2017-01-04
==================

  * Minor fix.
  * Phone number validation.
  * Fix: Do not update values in case of errors.
  * Polyfill `require.ensure` on node.
  * Validate zip code for US.
  * Kill getLangFromReq.
  * Fix return to previous location after login and increase auto-logout timeout to 15 minutes.
  * Implement business id.
  * Add country.
  * Add business website.
  * Fix FormMessages and related tests.
  * Revert dashboard and graphiql weired access. Just variables in .env for now.
  * Merge branch 'chore/change_email'
  * Implement change password.
  * Fix ignoring sensitive info in resolvers.
  * Animated Snackbar.
  * Add duration to Snackbar.
  * Cross-platform icons.
  * variables.scss -> _variables.scss
  * Access parse dashboard and graphiql more naturally.
  * Minor improvements.
  * Fix auto logout.
  * variables.scss -> _variables.scss
  * Access parse dashboard and graphiql more naturally.
  * Minor improvements.
  * Fix auto logout.
  * Don't send master key to the client.
  * Subscriptions initial stub.
  * Kill `onbeforeunload`. We will implement a `NavigationPrompt` later.
  * Small fix.
  * Centralized logging.
  * Small fix.
  * Cross-platform observers.
  * Use lightweight debounce.
  * [EXPERIMENTAL] Upgrade to webpack 2.
  * Clean schema.
  * Small fix.
  * Business registration, account settings and password change.
  * Fix a regression in `store` that prevented startedup.
  * Perform tests.
  * Remove express user-agent detection.
  * Node locale support.
  * Lint styles.
  * Fix notifications hidden.
  * Update deps.
  * Initial experimental support for theming.
  * Improve translations.
  * Make parse dashboard work.
  * Merge chunks.
  * Use single quotes consistently.
  * Display better errors messages...
  * Inline logo.
  * Better encapsulation.
  * One too many 'env'.
  * More renames.
  * Fix wrong naming.
  * Add support for linting graphql.
  * Fix comment.
  * Settle masterKey/sessionToken issues.
  * Just use redux's compose for now.
  * Update apollo-server.
  * User authentication initial commit.
  * Fix linting. Code now conforms to eslint config.
  * Cleanups.
  * Test resolver generators.
  * Setup initial test environment.
  * Minor tweaks.
  * Initiate readme.
  * Rename `BATCHING_INTERVAL` to `APOLLO_QUERY_BATCH_INTERVAL`.
  * Fix logging.
  * Load parse base on the execution environment: `parse/node` on server, `parse` in browser.

v0.0.1 / 2016-11-28
===================

  * Initial commit.


server/main.js
import './parse-config';

import 'isomorphic-fetch';

import getCurrentUser from 'getCurrentUser';

import users from 'data/_User';

import path from 'path';
import express from 'express';
import webpack from 'webpack';
import webpackConfig from 'build/webpack.config';
import config from 'build/config';
import compress from 'compression';

import cookie from 'react-cookie';
import createLocaleMiddleware from 'express-locale';
import cors from 'cors';

import createSSRRoute from './utils/createSSRRoute';

import { ParseServer } from 'parse-server';
import ParseDashboard from 'parse-dashboard';

import { graphqlExpress, graphiqlExpress } from 'graphql-server-express';

import schema from 'data/schema';

import { createServer } from 'http';
import { SubscriptionServer } from 'subscriptions-transport-ws';
import { subscriptionManager } from 'data/subscriptions';

// Connectors
import { UserConnector } from 'data/user/connector';
import { BusinessConnector } from 'data/business/connector';

// Models
import { Users } from 'data/user/models';
import { Business } from 'data/business/models';

import bodyParser from 'body-parser';

const log = require('log')('app:server');

const databaseUri = process.env.DATABASE_URI || process.env.MONGOLAB_URI;
if (!databaseUri) {
  log('DATABASE_URI not specified, falling back to localhost.');
}

const app = express();
const paths = config.utils_paths;

// Don't rate limit heroku
app.enable('trust proxy');

// Apply gzip compression
app.use(compress());

// Enable cors support
app.use(cors());

// Setup locale
app.use(createLocaleMiddleware());

// webpack compiler in devMode
const compiler = config.env === 'development' && webpack(webpackConfig);

// These are the current endpoints
// New endpoints must be added here in order for them to work.
const APP_PATHS = [
  '/',

  config.path_login,
  config.path_signup,
  config.path_choose_password,
  config.path_password_reset,

  config.path_email_verification_success,
  config.path_password_reset_success,
  config.path_invalid_link,

  // settings
  config.path_settings_base + '*',

  // product catalog
  config.path_product_catalog_base + '*',

  // search
  config.path_search,
];

if (config.ssrEnabled) {
  // Setup SSR
  log('Initializing server side rendering');
  const ssrRoute = createSSRRoute(app, compiler);
  app.get(APP_PATHS, ssrRoute);
} else {
  // This rewrites all routes requests to the root /index.html file
  // (ignoring file requests). If you want to implement universal
  // rendering, you'll want to remove this middleware.
  app.use(APP_PATHS, require('connect-history-api-fallback')({
    verbose : __DEV__,
    logger  : require('log')('app:server:history'),
  }));
}

// ------------------------------------
// Apply Webpack HMR Middleware
// ------------------------------------
if (config.env === 'development') {
  log('Enable webpack dev and HMR middleware');
  app.use(require('webpack-dev-middleware')(compiler, {
    publicPath  : webpackConfig.output.publicPath,
    contentBase : paths.client(),
    hot         : true,
    quiet       : config.compiler_quiet,
    noInfo      : config.compiler_quiet,
    lazy        : false,
    stats       : config.compiler_stats,
  }));
  app.use(require('webpack-hot-middleware')(compiler, {
    path: '/__webpack_hmr',
  }));

  // Serve static assets from ~/public since Webpack is unaware of
  // these files. This middleware doesn't need to be enabled outside
  // of development since this directory will be copied into ~/dist
  // when the application is compiled.
  app.use(express.static(paths.public()));
} else {
  log(
    'Server is being run outside of live development mode, meaning it will ' +
    'only serve the compiled application bundle in ~/dist. Generally you ' +
    'do not need an application server for this and can instead use a web ' +
    'server such as nginx to serve your static files. See the "deployment" ' +
    'section in the README for more information on deployment strategies.'
  );

  // Serving ~/dist by default. Ideally these files should be served by
  // the web server and not the app server, but this helps to demo the
  // server in production.
  if (config.serve_assets) {
    const options = {};

    if (config.ssrEnabled) {
      options.index = false;
    }
    app.use(express.static(paths.dist(), options));
  }
}

// ------------------------------------
// Parse server entrypoint
// ------------------------------------
if (!config.parse_database_uri) {
  log('DATABASE_URI not specified, falling back to localhost.');
}
const api = new ParseServer({
  appName                  : config.appName,
  databaseURI              : config.parse_database_uri || 'mongodb://localhost:27017/b2b-trading-platform',
  cloud                    : path.resolve(process.cwd(), 'backend', 'main.js'),
  appId                    : process.env.APPLICATION_ID,
  javascriptKey            : process.env.JAVASCRIPT_KEY,
  masterKey                : process.env.MASTER_KEY,
  serverURL                : config.parse_server_url || `http://${config.server_host}:${config.server_port}${config.parse_server_mount_point}`, // eslint-disable-line max-len
  publicServerURL                : config.parse_server_url || `http://${config.server_host}:${config.server_port}${config.parse_server_mount_point}`, // eslint-disable-line max-len
  enableAnonymousUsers     : process.env.ANON_USERS === 'yes',
  allowClientClassCreation : true,
  maxUploadSize            : '25mb',

  // Retricts sesssions to 15 days
  sessionLength            : 15 * 24 * 60 * 60,

  // Email
  verifyUserEmails                 : true,
  revokeSessionOnPasswordReset     : true,
  emailVerifyTokenValidityDuration : 2 * 24 * 60 * 60, // 2 days
  emailAdapter                     : {
    module: 'parse-server-mailgun',
    options: {
      fromAddress: config.mailgun_from_address,
      domain: config.mailgun_domain,
      apiKey: config.mailgun_api_key,
      templates: {
        passwordResetEmail: {
          subject: 'Reset your password for ' + config.appName,
          pathPlainText: paths.server('email-templates/password_reset_email.txt'),
          callback: (user) => ({}),
        },
        verificationEmail: {
          subject: 'Please verify your e-mail for ' + config.appName,
          pathPlainText: paths.server('email-templates/verification_email.txt'),
          callback: (user) => ({}),
        },
      },
    },
  },

  passwordPolicy : {
    doNotAllowUsername         : true,
    resetTokenValidityDuration : 2 * 60 * 60, // 2 hours
  },

  userSensitiveFields: [
  ],

  customPages: {
    invalidLink          : `http://${config.server_host}:${config.server_port}` + config.path_invalid_link,
    verifyEmailSuccess   : `http://${config.server_host}:${config.server_port}` + config.path_email_verification_success,
    choosePassword       : `http://${config.server_host}:${config.server_port}` + config.path_choose_password,
    passwordResetSuccess : `http://${config.server_host}:${config.server_port}` + config.path_password_reset_success,
  },
});

// Serve the Parse API on the /parse URL prefix
app.use(config.parse_server_mount_point, api);

// ------------------------------------
// Parse dashboard entrypoint
// ------------------------------------
const dashboard = new ParseDashboard({
  'apps': [
    {
      'appId'         : process.env.APPLICATION_ID,
      'javascriptKey' : process.env.JAVASCRIPT_KEY,
      'masterKey'     : process.env.MASTER_KEY,
      'serverURL'     : config.parse_server_url || `http://${config.server_host}:${config.server_port}${config.parse_server_mount_point}`,  // eslint-disable-line max-len
      'appName'       : config.appName,
      'production'    : !__DEV__,
    },
  ],
  'trustProxy': 1,
  'useEncryptedPasswords': true,
  'users': users.map(function (user) {
    return {
      'user' : user.username,
      'pass' : user.bcryptPassword,
    };
  }),
}, /* allowInsecureHTTP = */ __DEV__);

// Serve dashboard
app.use(config.parse_dashboard_mount_point, dashboard);

// ------------------------------------
// Graphql server entrypoint
// ------------------------------------
app.use(config.graphql_endpoint, bodyParser.json(), graphqlExpress((req, res) => {
  // Get the query, the same way express-graphql does it
  // https://github.com/graphql/express-graphql/blob/3fa6e68582d6d933d37fa9e841da5d2aa39261cd/src/index.js#L257
  const query = req.query.query || req.body.query;
  if (query && query.length > 2000) {
    // None of our app's queries are this long
    // Probably indicates someone trying to send an overly expensive query
    throw new Error('Query too large.');
  }

  cookie.plugToRequest(req, res);

  const user = getCurrentUser();
  return {
    schema,
    context: {
      user,
      Users: new Users({ user, connector: new UserConnector() }),
      Business: new Business({ user, connector: new BusinessConnector() }),
    },
    logFunction: require('log')('app:server:graphql'),
    debug: __DEV__,
  };
}));

if (__DEV__) {
  app.use(config.graphiql_endpoint, graphiqlExpress({
    endpointURL: config.graphql_endpoint,
  }));
}

// WebSocket server for subscriptions
const websocketServer = createServer((request, response) => {
  response.writeHead(404);
  response.end();
});

// eslint-disable-next-line
new SubscriptionServer(
  {
    subscriptionManager,

    // the obSubscribe function is called for every new subscription
    // and we use it to set the GraphQL context for this subscription
    onSubscribe: (msg, params) => {
      return Object.assign({}, params, {
        context: {
          Business: new Business({ connector: new BusinessConnector() }),
        },
      });
    },
  },
  websocketServer
);
module.exports = { app, websocketServer };

